### PR TITLE
Refactoring and runtime fixes for junkie and smoker

### PR DESCRIPTION
**Pull request by actioninja**
I just accidentally pushed this to master, shit

## About The Pull Request

Some refactors and fixes for the junkie quirk.
I don't know why this trash got merged this code was awful, I wonder what idiot wrote it.
image
oh

### CODE REVIEW

**Comment by nemvar** [comment]
This doesn't fix the main runtimes. Needs to create it's references in add() instead of in on_spawn()

**Comment by actioninja** [comment]
from my testing this does fix it, even if it's a somewhat shitty solution for it.

**Comment by optimumtact** [comment]
these runtiems are the worse

File: code/datums/traits/negative.dm [hunk 1]
```diff
@@ -496,19 +493,13 @@
...
-	if (tick_counter == 60) //Halfassed optimization, increase this if there's slowdown due to this quirk
-		var/in_list = FALSE
-		for (var/datum/reagent/entry in H.reagents.addiction_list)
-			if(istype(entry, reagent_type))
-				in_list = TRUE
-				break
-		if(!in_list)
+	if(world.time > next_process)
+		next_process = world.time + process_interval
+		if(!H?.reagents?.addiction_list?.Find(reagent_instance))
```
Comment (AnturK) [inline code/datums/traits/negative.dm]: These ?. are completely redundant if you access H.reagents.addiction_list right after.

**Review by AnturK — CHANGES_REQUESTED** [review]
Not sure what this fixes but anyway

File: code/datums/traits/negative.dm (hunk 1 above)
Comment (actioninja) [reply code/datums/traits/negative.dm]: Yeah, good point, it was just some overzealous safety checking that didn't actually have to do with the runtime.

Comment (actioninja) [reply code/datums/traits/negative.dm]: the runtime is on the lines below, sometime reagent_instance gets nulled out (cloning?) and I'm not sure why exactly. This just creates a new instance if it's nulled.

**Pull request merged**